From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: LexManos <LexManos@gmail.com>
Date: Wed, 10 Jun 2020 09:59:19 -0700
Subject: [PATCH] Add -only argument It will filter what classes are decompiled
 from the target jar.

Uses a prefix system, so -only=net/minecraft/block/ will decompile all classes in the block package and subpackages.
Useful for debugging to limit scope/runtime.

diff --git a/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java b/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
index 8aefba318b0f09b559d41d945a3d8099a861ac55..305b06903b36bc4cdb740a8f050543403e6406fe 100644
--- a/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
+++ b/src/org/jetbrains/java/decompiler/main/ClassesProcessor.java
@@ -38,6 +38,7 @@ public class ClassesProcessor implements CodeConstants {
 
   private final StructContext context;
   private final Map<String, ClassNode> mapRootClasses = new HashMap<>();
+  private final Set<String> whitelist = new HashSet<>();
 
   private static class Inner {
     private String simpleName;
@@ -63,6 +64,22 @@ public class ClassesProcessor implements CodeConstants {
     this.context = context;
   }
 
+  public void addWhitelist(String prefix) {
+    this.whitelist.add(prefix);
+  }
+
+  public boolean isWhitelisted(String cls) {
+    if (this.whitelist.isEmpty())
+      return true;
+
+    for (String prefix : this.whitelist) {
+      if (cls.startsWith(prefix))
+        return true;
+    }
+
+    return false;
+  }
+
   public void loadClasses(IIdentifierRenamer renamer) {
     Map<String, Inner> mapInnerClasses = new HashMap<>();
     Map<String, Set<String>> mapNestedClassReferences = new HashMap<>();
@@ -162,9 +179,11 @@ public class ClassesProcessor implements CodeConstants {
           }
         }
 
-        ClassNode node = new ClassNode(ClassNode.CLASS_ROOT, cl);
-        node.access = cl.getAccessFlags();
-        mapRootClasses.put(cl.qualifiedName, node);
+        if (isWhitelisted(cl.qualifiedName)) {
+          ClassNode node = new ClassNode(ClassNode.CLASS_ROOT, cl);
+          node.access = cl.getAccessFlags();
+          mapRootClasses.put(cl.qualifiedName, node);
+        }
       }
       linkEnclosingMethods(cl);
     }
diff --git a/src/org/jetbrains/java/decompiler/main/Fernflower.java b/src/org/jetbrains/java/decompiler/main/Fernflower.java
index ab8c1a60acfdef8c6ffb3082efbfbc1315d2c0bc..469b313a958cf8aca35fe98fc809880d12db671e 100644
--- a/src/org/jetbrains/java/decompiler/main/Fernflower.java
+++ b/src/org/jetbrains/java/decompiler/main/Fernflower.java
@@ -115,6 +115,10 @@ public class Fernflower implements IDecompiledData {
     structContext.saveContext();
   }
 
+  public void addWhitelist(String prefix) {
+    classProcessor.addWhitelist(prefix);
+  }
+
   public void clearContext() {
     DecompilerContext.setCurrentContext(null);
   }
@@ -122,7 +126,7 @@ public class Fernflower implements IDecompiledData {
   @Override
   public String getClassEntryName(StructClass cl, String entryName) {
     ClassNode node = classProcessor.getMapRootClasses().get(cl.qualifiedName);
-    if (node.type != ClassNode.CLASS_ROOT) {
+    if (node == null || node.type != ClassNode.CLASS_ROOT) {
       return null;
     }
     else if (converter != null) {
diff --git a/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java b/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
index 5393ddec202efa7e1f6bb3f7740fa83a70ea48fd..2b6e9ed5d2cf8e3781c9b33b1489fee9447d8181 100644
--- a/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
+++ b/src/org/jetbrains/java/decompiler/main/decompiler/ConsoleDecompiler.java
@@ -66,6 +66,7 @@ public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
     Map<String, Object> mapOptions = new HashMap<>();
     List<File> sources = new ArrayList<>();
     List<File> libraries = new ArrayList<>();
+    Set<String> whitelist = new HashSet<>();
 
     boolean isOption = true;
     for (int i = 0; i < args.length - 1; ++i) { // last parameter - destination
@@ -88,6 +89,9 @@ public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
         if (arg.startsWith("-e=")) {
           addPath(libraries, arg.substring(3));
         }
+        else if (arg.startsWith("-only=")) {
+          whitelist.add(arg.substring(6));
+        }
         else {
           addPath(sources, arg);
         }
@@ -114,6 +118,9 @@ public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
     for (File source : sources) {
       decompiler.addSource(source);
     }
+    for (String prefix : whitelist) {
+      decompiler.addWhitelist(prefix);
+    }
 
     decompiler.decompileContext();
   }
@@ -151,6 +158,10 @@ public class ConsoleDecompiler implements IBytecodeProvider, IResultSaver {
     engine.addLibrary(library);
   }
 
+  public void addWhitelist(String prefix) {
+    engine.addWhitelist(prefix);
+  }
+
   public void decompileContext() {
     try {
       engine.decompileContext();
